cmd/panfs-multipart.go

```diff
@@ -698,36 +796,45 @@ func (fs *PANFSObjects) CompleteMultipartUpload(ctx context.Context, bucket stri
 	}
 
 	appendFallback := true // In case background-append did not append the required parts.
-	appendFilePath := pathJoin(bucketPath, panfsS3TmpDir, fs.nodeDataSerial, bgAppendsDirName, fmt.Sprintf("%s.%s", uploadID, mustGetUUID()))
+	appendFilePath := fs.getObjectBackgroundAppendPath(bucketPath, object, uploadID)
 
 	// Most of the times appendFile would already be fully appended by now. We call fs.backgroundAppend()
 	// to take care of the following corner case:
 	// 1. The last PutObjectPart triggers go-routine fs.backgroundAppend, this go-routine has not started yet.
 	// 2. Now CompleteMultipartUpload gets called which sees that lastPart is not appended and starts appending
 	//    from the beginning
-	fs.backgroundAppend(ctx, bucket, object, uploadID)
-
+	// there are no info about mpart upload in memory
 	fs.appendFileMapMu.Lock()
 	file := fs.appendFileMap[uploadID]
-	delete(fs.appendFileMap, uploadID)
+	if file == nil {
+		// This may happen when complete upload is called from node where no parts were uploaded
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
```

Comment:
actually, I would make cleanup of fs.appendFileMap as last step and let me describe why. In case of parallel complete (or abort) requests for single object (that may be happen due to application error), we may create two separate instances of _panfsAppendFile_ structures. Why is this bad? We will have two lock file instances with separate file descriptors and local mutexes that protect them. This is very bad as it can break lockfile logic.

What I'd suggest, is a two-step cleanup of that map and the following rules for entries in appendFileMap:

1) we can add e.g. boolean flag e.g. 'stale' to  _panfsAppendFile_ structure, which is false by default
2) every time when abort/complete/background append/put part/etc. read _fs.appendFileMap_ and find existing entry, they check 'stale' flag and if it's 'false', they just assume that abort/complete is in progress and return some kind of error
3) complete and abort clean up the entry in map in two steps:

```
// step 1:
    lock map
    file := fs.appendFileMap[uploadID]
    if file == nil {
        file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
        fs.appendFileMap[uploadID] = file
    } else if file.stale {
        // abort as another complete/abort is in progress
    }
    file.stale = true
    unlock map

// Main work
    lock file.flock
    do all work

// step 2:
    lock map
    delete(map, uploadID)
    unlock map
```

Let me know if I missed something.